### PR TITLE
fix(web): first-draft, usable build for module-based app/browser 🧩

### PR DESCRIPTION
**Pull request by jahorton**
At long last... we can finally spin up a full-body instance of Keyman Engine for Web that's 100% module-based!

Of course, this version is missing several components - not all of the codebase is modularized yet, after all.  Even still, this restores KMW with a lot of its baseline functionality in place:
- Page focus & its relationship to context management are generally as expected
- Both hardware-keyboard and on-screen-keyboard input operate properly
- The touch form-factor language menu properly switches keyboards
- The desktop OSK's persistent-state storage operates as expected.

Some very notable caveats, however:
- There is currently no functional desktop form-factor language menu.  None of the UI modules (outside of 'touch') are operational.
- Attachment fonts are not currently operational; they'll be restored in a follow-up.  
    - Handled in #8820.
- Hotkeys and device-rotation handlers are non-operational.
- Much of KMW's published API is missing.
- Only one testing page has been restored at this point:  src/test/manual/unminified.html.

----

There are a fair few bug fixes among the changeset for this PR; these were necessary to restore aspects of our basic, expected functionality and/or to prevent the engine from crashing during initialization or operation.  I've pulled a number of fixes forward into earlier PRs as reasonable, but some of the fixes didn't fit well anywhere else... so I've left them here.

Due to the number of remaining caveats, I will not be specifying user testing at this time.

@keymanapp-test-bot skip

### CODE REVIEW

File: web/README.md [hunk 1]
```diff
@@ -64,3 +64,73 @@ the former command executes.
    tab.
 4. In the Dev console, you can set a breakpoint in your test and refresh the
    page to debug
+
+### Approximate Overall Design
+
+```mermaid
+graph TD;
```
Comment (jahorton) [inline web/README.md:71]: While this code block below doesn't render well offline, or in reviews, it displays beautifully when directly viewing it within GitHub.

We should be able to prerender it for offline viewing using https://github.com/mermaid-js/mermaid-cli if desired; that does mean adding more dev-dependencies, though.

File: web/src/app/browser/src/contextManager.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@ interface KeyboardCookie {
  * @param       {Object}      Ptarg      Target element
  */
 function _SetTargDir(Ptarg: HTMLElement, activeKeyboard: Keyboard) {
-  var elDir=(activeKeyboard && activeKeyboard.isRTL) ? 'rtl' : 'ltr';
+  var elDir=(activeKeyboard && activeKeyboard?.isRTL) ? 'rtl' : 'ltr';
```
Comment (mcdurdin) [inline web/src/app/browser/src/contextManager.ts:29]:
```suggestion
  const elDir = activeKeyboard?.isRTL ? 'rtl' : 'ltr';
```

File: web/src/app/browser/src/keymanEngine.ts [hunk 3]
```diff
@@ -30,12 +34,15 @@ export class KeymanEngine extends KeymanEngineBase<BrowserConfiguration, Context
...
+      const e = target?.getElement();
+      (this.osk.activationModel as TwoStateActivator<HTMLElement>).activationTrigger = e;
+
       if(this.config.hostDevice.touchable) {
         if(!target || !this.osk) {
           return;
         }
 
-        const e = target.getElement();
+        const e = target?.getElement();
```
Comment (mcdurdin) [inline web/src/app/browser/src/keymanEngine.ts:45]:
```suggestion
```

See l.37

File: web/src/app/browser/src/keymanEngine.ts [hunk 4]
```diff
@@ -30,12 +34,15 @@ export class KeymanEngine extends KeymanEngineBase<BrowserConfiguration, Context
 
     // Scrolls the document-body to ensure that a focused element remains visible after the OSK appears.
     this.contextManager.on('targetchange', (target: OutputTarget<any>) => {
+      const e = target?.getElement();
```
Comment (mcdurdin) [inline web/src/app/browser/src/keymanEngine.ts:37]: Need to consider when `e` is nullish, e.g. see l.47/54 where `e` is used.

File: web/src/app/browser/src/keymanEngine.ts [hunk 5]
```diff
@@ -30,12 +34,15 @@ export class KeymanEngine extends KeymanEngineBase<BrowserConfiguration, Context
...
+      (this.osk.activationModel as TwoStateActivator<HTMLElement>).activationTrigger = e;
+
       if(this.config.hostDevice.touchable) {
         if(!target || !this.osk) {
```
Comment (mcdurdin) [inline web/src/app/browser/src/keymanEngine.ts:41]:
```suggestion
        if(!target || !e || !this.osk) {
```

File: web/src/app/browser/src/keymanEngine.ts [hunk 6]
```diff
@@ -164,6 +188,51 @@ export class KeymanEngine extends KeymanEngineBase<BrowserConfiguration, Context
...
+      });
+    } else {
+      let x: (string|KeyboardStub)[] = [];
+      if (Array.isArray(args[0])) {
+        args[0].forEach(a =>
+          x.push(a));
+      } else if (Array.isArray(args)) {
+        args.forEach(a =>
+          x.push(a));
+      } else {
```
Comment (mcdurdin) [inline web/src/app/browser/src/keymanEngine.ts:214]:
```suggestion
      if (Array.isArray(args[0])) {
        x.push(...args[0]);
      } else if (Array.isArray(args)) {
        x.push(...args);
      } else {
```

**Review by mcdurdin — APPROVED** [review]

File: web/src/app/browser/src/utilApiEndpoint.ts
Comment (jahorton) [inline web/src/app/browser/src/utilApiEndpoint.ts:1]: Whoops!  This fellow slipped in far too early; I'm guessing I accidentally pulled it forward in a rebase at some point.

It happened in #8813, based on the `git blame` report.

**Pull request merged**